### PR TITLE
BUG: fixed issue where Index was being accessed with DatetimeIndex attribute

### LINKED ISSUE #49925: BUG: Holiday.dates called with date range outside of Holiday's `start_date/end_date` raises AttributeError when Holiday has `observance` and `days_of_week`
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
from pandas.tseries.holiday import Holiday, next_monday
hol = Holiday("Dec 25", month=12, day=25, observance=next_monday, days_of_week=(5, 6), end_date=pd.Timestamp("2013-01-01"))
hol.dates(pd.Timestamp("2020-01-01"), pd.Timestamp("2023-01-01"))
```

### Issue Description

When a Holiday has an `observance`, `days_of_week `and a `start_date/end_date` where the holiday applies, if you run the `dates` method for a date range outside of the holiday's `start_date/end_date` you get an error. The problem arises because [`self.apply_rules(dates)` returns an `Index` instead of a `DatetimeIndex` when it applies the observance over `dates` (an empty `DatetimeIndex`)](https://github.com/pandas-dev/pandas/blob/v1.5.2/pandas/tseries/holiday.py#L338-L339). If `self.dates_of_week` is not None[ `self.dates` tries to access the `dayofweek` attribute](https://github.com/pandas-dev/pandas/blob/v1.5.2/pandas/tseries/holiday.py#L272-L274) on an `Index` object which doesn't exist.

```
Python 3.11.0 | packaged by conda-forge | (main, Oct 25 2022, 06:12:32) [MSC v.1929 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pandas as pd
>>> from pandas.tseries.holiday import Holiday, next_monday
>>> hol = Holiday("Dec 25", month=12, day=25, observance=next_monday, days_of_week=(5, 6), end_date=pd.Timestamp("2013-01-01"))
>>> hol.dates(pd.Timestamp("2020-01-01"), pd.Timestamp("2023-01-01"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\marco\anaconda3\envs\pandas\Lib\site-packages\pandas\tseries\holiday.py", line 274, in dates
    np.in1d(holiday_dates.dayofweek, self.days_of_week)
            ^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'Index' object has no attribute 'dayofweek'
```

### Expected Behavior

It should return an empty `DatetimeIndex`, of the same dtype as `self._reference_dates`
```
>>>import pandas as pd
>>>from pandas.tseries.holiday import Holiday, next_monday
>>>hol = Holiday("Dec 25", month=12, day=25, observance=next_monday, days_of_week=(5, 6), end_date=pd.Timestamp("2013-01-01"))
>>>hol.dates(pd.Timestamp("2020-01-01"), pd.Timestamp("2023-01-01"))
DatetimeIndex([], dtype='datetime64[ns]', freq='<DateOffset: years=1>')
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 8dab54d6573f7186ff0c3b6364d5e4dd635ff3e7
python           : 3.11.0.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.22621
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_Australia.1252

pandas           : 1.5.2
numpy            : 1.23.5
pytz             : 2022.6
dateutil         : 2.8.2
setuptools       : 65.5.1
pip              : 22.3.1
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None
tzdata           : None

</details>

### PR DESCRIPTION
**Pull request by bilaljai**
- [x] Closes #49925 
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Changes: updated the dates method in pandas/tseries/holiday.py to include a check to prevent the DatetimeIndex attribute
from being accessed by an Index object.  The bug occurs if the holidays land outside the provided date range which results in an empty DatetimeIndex object being converted to an Index object. The updated dates method now checks if the object contains elements before allowing access to a DatetimeIndex attribute because such an object will be of type Index. Furthermore, the Index object is converted back into a DatetimeIndex object. A test was added to pandas/tests/tseries/holiday/test_holiday.py to ensure that the intended behavior is occurring.

### CODE REVIEW

**Review by phofl — COMMENTED** [review]
Can you run pre-commit? Your formatting is off

File: doc/source/whatsnew/v2.0.0.rst [hunk 1]
```diff
@@ -178,6 +178,37 @@ the appropriate rows. Also, when ``n`` is larger than the group, no rows instead
 .. ipython:: python
 
     gb.nth(n=3, dropna="any")
+    
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In previous versions, an error occured if the :meth:`dates` method is ran for a date range
```
[the remark this answers is not in the document]
Comment (bilaljai) [reply doc/source/whatsnew/v2.0.0.rst]: Thanks for the feedback I've ran pre-commit and updated v2.0.0.rst to include your suggestion.

File: doc/source/whatsnew/v2.0.0.rst [hunk 2]
```diff
@@ -179,6 +179,10 @@ the appropriate rows. Also, when ``n`` is larger than the group, no rows instead
 
     gb.nth(n=3, dropna="any")
 
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In previous versions, an error occurred if the :meth:`dates` method is ran for a date range.
```
Comment (phofl) [inline doc/source/whatsnew/v2.0.0.rst:184]: You can just add this to one of the bug fixes sections below

File: pandas/tests/tseries/holiday/test_holiday.py [hunk 3]
```diff
@@ -193,6 +193,13 @@ def test_holidays_within_dates(holiday, start, expected):
         holiday.dates(utc.localize(Timestamp(start)), utc.localize(Timestamp(start)))
     ) == [utc.localize(dt) for dt in expected]
 
+    
+def test_holidays_outside_dates():
+    hol = Holiday("Dec 25", month=12, day=25, observance=next_monday, days_of_week=(5, 6),
+                  end_date=Timestamp("2013-01-01"))
```
Comment (phofl) [inline pandas/tests/tseries/holiday/test_holiday.py:199]: The formatting is off, pre-commit should fix this

File: doc/source/whatsnew/v2.0.0.rst (hunk 2 above)
Comment (bilaljai) [reply doc/source/whatsnew/v2.0.0.rst:184]: Ah I see I'll work on it now

Comment (bilaljai) [reply doc/source/whatsnew/v2.0.0.rst]: I've added this to the Timezones section.

File: pandas/tests/tseries/holiday/test_holiday.py (hunk 3 above)
Comment (bilaljai) [reply pandas/tests/tseries/holiday/test_holiday.py]: Ran pre-commit, formatting fixed now.

File: doc/source/whatsnew/v2.0.0.rst [hunk 4]
```diff
@@ -717,7 +717,7 @@ Timezones
 ^^^^^^^^^
 - Bug in :meth:`Series.astype` and :meth:`DataFrame.astype` with object-dtype containing multiple timezone-aware ``datetime`` objects with heterogeneous timezones to a :class:`DatetimeTZDtype` incorrectly raising (:issue:`32581`)
 - Bug in :func:`to_datetime` was failing to parse date strings with timezone name when ``format`` was specified with ``%Z`` (:issue:`49748`)
--
+- Bug in :meth:`dates` was trying to access an DatetimeIndex attribute on an Index object when holiday landed outside of date range provided (:issue:`49925`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v2.0.0.rst:720]: can you write out the full class.method name, otherwise not clear what "dates" refers to

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**